libimage/filters.go

```diff
@@ -272,55 +278,100 @@ func filterManifest(ctx context.Context, value bool) filterFunc {
 	}
 }
 
-// filterReferences creates a reference filter for matching the specified value.
-func filterReferences(r *Runtime, value string) filterFunc {
-	lookedUp, _, _ := r.LookupImage(value, nil)
+// filterReferences creates a reference filter for matching the specified wantedReferenceMatches value (OR logic)
+// and for matching the unwantedReferenceMatches values (AND logic)
+func filterReferences(r *Runtime, wantedReferenceMatches, unwantedReferenceMatches []string) filterFunc {
 	return func(img *Image) (bool, error) {
-		if lookedUp != nil {
-			if lookedUp.ID() == img.ID() {
+		// Empty reference filters, return true
+		if len(wantedReferenceMatches) == 0 && len(unwantedReferenceMatches) == 0 {
+			return true, nil
+		}
+
+		unwantedMatched := false
+		// Go through the unwanted matches first
+		for _, value := range unwantedReferenceMatches {
+			matches, err := imageMatchesReferenceFilter(r, img, value)
+			if err != nil {
+				return false, err
+			}
+			if matches {
+				unwantedMatched = true
+			}
+		}
+
+		// If there are no wanted match filters, then return false for the image
+		// that matched the unwanted value otherwise return true
+		if len(wantedReferenceMatches) == 0 {
+			if unwantedMatched {
+				return false, nil
+			}
+			return true, nil
+		}
+
+		// Go through the wanted matches
+		// If there are unwanted and wanted matches together, the output will be based on
+		// the wanted filters
```

Comment:
That would make any use of both `wanted…` and `unwanted…` at the same time a clear user error.

So, either that situation should be detected and rejected outright (and `filterReferences` then can return either a wanted-only or an unwanted-only lambda, and those lambdas can be simpler), or actually allowed and supported.

Because the matches use wildcards, using `unwanted` together with` wanted` is not _obviously redundant_ (in something vaguely like `wanted=(prefix1*, prefix2*), unwanted=(*suffix1, *suffix2)`); also it was allowed previously (although we are changing the semantics). So I’m weakly leaning towards “allowed and supported”.

Reply:
Fixed to match current behaviour and be allowed and supported